src/types/ianatimezone.jl

```diff
@@ -0,0 +1,159 @@
+const IANA_TABLE_SIZE = 1680
+
+perfect_hash(tz::VariableTimeZone) = perfect_hash(tz.name)
+function perfect_hash(str::AbstractString)
+    # This was generated via `gperf` and translated by hand.
+    # in case of collisions from new keys being added you must *not* regenerate it or
+    # change the assoc table, as that will change existing hashes and break any serialized
+    # data. Instead add special cases from the exact string via adding special cases. e.g.
+    #`name=="New/Timezone" && return 1681` (after adjusting `IANA_TABLE_SIZE`)
+    # The code below will (for all timezones in the 2021a release of tzdata) generate a
+    # value between 29 and 1680 inclusive. So any new keys that need to be added
+    # manually  to resolve collisions can freely use anything outside that range.
+
+    asso_values = (
+      1681, 1681, 1681, 1681, 1681, 1681, 1681, 1681, 1681,
+      1681, 1681, 1681, 1681, 1681, 1681, 1681, 1681, 1681, 1681,
+      1681, 1681, 1681, 1681, 1681, 1681, 1681, 1681, 1681, 1681,
+      1681, 1681, 1681, 1681, 1681, 1681, 1681, 1681, 1681, 1681,
+      1681, 1681, 1681,    4, 1681,    1,    3,    3,  241,   48,
+         1,   22,   10,    9,   61,   18,   74,    9,   30, 1681,
+      1681, 1681, 1681, 1681, 1681,    4,    2,   18,  167,   48,
+       547,  691,  532,  446,  574,  466,  211,  452,    8,  387,
+       110,  475,  601,  262,   83,  124,  580,  536,   60,    1,
+       719,    2, 1681, 1681, 1681,  291,    1,    1,   27,  424,
+        13,    5,   43,    3,  355,   12,  168,  148,   90,  179,
+         4,    1,  315,    3,    3,    3,    2,   37,    5,   65,
+        22,  320,  245,    1, 1681, 1681, 1681,
+    )
+
+    units = codeunits(str)
+    len = length(units)
+    hval = len
+
+    len >= 19 && (hval += asso_values[units[19]])
+    len >= 12 && (hval += asso_values[units[12]])
+    len >= 11 && (hval += asso_values[units[11]])
+    len >= 9 && (hval += asso_values[units[9] + 1])
+    len >= 8 && (hval += asso_values[units[8]])
+    len >= 6 && (hval += asso_values[units[6] + 1])
+    len >= 4 && (hval += asso_values[units[4]])
+    len >= 2 && (hval += asso_values[units[2] + 1])
+    len >= 1 && (hval += asso_values[units[1]])
+    len > 0 && (hval += asso_values[units[end]])  # add the last
```

Comment:
```suggestion
    @inbounds begin
        len >= 19 && (hval += asso_values[units[19]])
        len >= 12 && (hval += asso_values[units[12]])
        len >= 11 && (hval += asso_values[units[11]])
        len >= 9 && (hval += asso_values[units[9] + 1])
        len >= 8 && (hval += asso_values[units[8]])
        len >= 6 && (hval += asso_values[units[6] + 1])
        len >= 4 && (hval += asso_values[units[4]])
        len >= 2 && (hval += asso_values[units[2] + 1])
        len >= 1 && (hval += asso_values[units[1]])
        len > 0 && (hval += asso_values[units[end]])  # add the last
    end
```
This saves some time and simplifies the native code but relies on each code unit being in the range `1:125`. 

Might be worth it just to pad the array of `asso_values` to cover the domain of `UInt8`

Reply:
> This saves some time and simplifies the native code but relies on each code unit being in the range 1:125

I tried this and didn't notice any improvement.
How were you benchmarking it?

> Might be worth it just to pad the array of asso_values to cover the domain of UInt8

I tried this and it made it *way* worse. I think because the tuple gets too lange for julia to be willing to do some optimizations.

Reply:
```
julia> @btime for x in ["America/Winnipeg", "Etc/UTC", "Canada/Central"]
           perfect_hash(x)
       end
  71.438 ns (1 allocation: 112 bytes)

julia> @btime for x in ["America/Winnipeg", "Etc/UTC", "Canada/Central"]
           perfect_hash_inbounds(x)
       end
  66.137 ns (1 allocation: 112 bytes)
```

Reply:
running `@btime` on the whole testset in `test/types/ianatimezone.jl`
With the benchmark

```julia
@btime for x in ("America/Winnipeg", "Etc/UTC", "Canada/Central")
                TimeZones.perfect_hash(x)
             end
```
Noting that i got rid of the vector in the benchmark code since that was dominating.
In julia 1.6

 - current code without the inbounds annotation|:   38.097 ns (0 allocations: 0 bytes)
 - with inbounds: 33.217 ns (0 allocations: 0 bytes)
 - with inbounds + make assoc_values bigger: 79.305 ns (0 allocations: 0 bytes)
 - with inbounds + prechecking: 33.465 ns (0 allocations: 0 bytes)
 
 
 Looks like we have a winner with the prechecking since that is safe.
 I will push that shortly

Reply:
- with inbounds and assoc to 127: 23.236 ns (0 allocations: 0 bytes) (Not safe)
 - with inbounds and checking based on bitmask: 47.035 ns (0 allocations: 0 bytes)
 - with inbounds and checking based in <: 36.359 ns (0 allocations: 0 bytes)
 - with inbounds and assoc to 255: 44.585 ns (0 allocations: 0 bytes)

Reply:
Hmm i don't trust my benchmarks
Timing for 127 + checking based on <: sometimes is 36ns sometimes is 48ns.

I think we should do that though, and i will push that code